### PR TITLE
Feature/custom_entity

### PR DESCRIPTION
**Pull request by Pho3niX90**
The new CustomEntity interface allows for streamlined handling of state-related tasks and boilerplate code reduction. Additionally, the full battery capacity computation logic is now encapsulated within the `getBatteryCapacity` method. This promotes better organization and clarity throughout the codebase.

I added the co-authored tag. Although I "think" I got all the logic, I might not have understood the intent at all of the calculations, so just need another pair of eyes.

### CODE REVIEW

File: src/inverters/brands/huawei.ts [hunk 1]
```diff
@@ -1,6 +1,7 @@
 import {InverterSettingsDto, InverterStatus} from '../dto/inverter-settings.dto';
 import {InverterModel} from '../../types';
 import {localize} from '../../localize/localize';
+import {Utils} from '../../helpers/utils';
```
Comment (slipx06) [inline src/inverters/brands/huawei.ts:4]: What was the reason for importing this?

Comment (Pho3niX90) [reply src/inverters/brands/huawei.ts:4]: I think it's unnecessary code. 

The new battery function on the base class was extended here to overwrite the function. However, I left it out for now and left the functionality in the index until I can test it properly. 

So this import is left over/unnecessary

File: src/index.ts [hunk 2]
```diff
@@ -127,9 +127,9 @@ export class SunsynkPowerFlowCard extends LitElement {
         const state_day_aux_energy = this.getEntity('day_aux_energy');
 
         //Inverter
-        const state_inverter_voltage = this.getEntity('inverter_voltage_154', {state: '0'});
-        const state_load_frequency = this.getEntity('load_frequency_192', {state: '0'});
-        const state_inverter_current = this.getEntity('inverter_current_164', {state: '0'});
+        const state_inverter_voltage = this.getEntity('inverter_voltage_154');
```
Comment (slipx06) [inline src/index.ts:130]: The getEntity function defaults to a state of 0, is the the reason for tremoving the defult value {state: '0'}. If yes can we remove it everywhere?

File: src/types.ts [hunk 3]
```diff
@@ -36,6 +36,7 @@ export enum AutarkyType {
 
 export interface sunsynkPowerFlowCardConfig extends LovelaceCardConfig {
     type: string;
+    dynamicConfig: boolean;
```
Comment (slipx06) [inline src/types.ts:39]: Is this needed?

File: src/index.ts (hunk 2 above)
Comment (Pho3niX90) [reply src/index.ts:130]: Yes you are correct, since the method already defaults tk that, it was removed.

File: src/types.ts (hunk 3 above)
Comment (Pho3niX90) [reply src/types.ts:39]: Definitely not. Wanted to add the functionality for the dynamic hiding of elements, but just seems to much maintenance for too little gain at this stage.

File: src/inverters/dto/custom-entity.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
+import {HassEntity} from 'home-assistant-js-websocket/dist/types';
+import {Utils} from '../../helpers/utils';
+
+/**
+ * CustomEntity interface represents a custom entity in Home Assistant.
```
Comment (slipx06) [inline src/inverters/dto/custom-entity.ts:5]: This is great. Thanks

File: src/index.ts [hunk 5]
```diff
@@ -785,8 +730,8 @@ export class SunsynkPowerFlowCard extends LitElement {
         let max_linewidth = (Utils.toNum(config.max_line_width) < 1 ? 1 : config.max_line_width) - 1;
         let min_linewidth = Utils.toNum(config.min_line_width) || 1;
 
-        const BatteryMaxPower = this.hass.states[config.battery.max_power] || {state: config.battery.max_power ?? ''};
-        let BattMaxPower = Utils.toNum(BatteryMaxPower.state);
+        const BatteryMaxPower = this.getEntity('max_power', {state: config.battery.max_power ?? ''});
```
Comment (slipx06) [inline src/index.ts:733]: Will this work. max_power is defined under multiple config sections i.e. solar, battery etc. How do we ensure it matches on battery.max_power

Comment (Pho3niX90) [reply src/index.ts]: I made a change to better access the relevant objects property.

**Comment by slipx06** [comment]
I'm getting this error

image

File: src/types.ts [hunk 6]
```diff
@@ -57,16 +57,16 @@ export interface sunsynkPowerFlowCardConfig extends LovelaceCardConfig {
         three_phase: boolean;
     }
     battery: {
-        energy: number;
-        shutdown_soc: number;
-        shutdown_soc_offgrid: number;
+        energy: any;
```
Comment (slipx06) [inline src/types.ts:60]: Please confirm that I am corrct to change this to `any` where you can provide a number or sensor

**Comment by Pho3niX90** [comment]
@slipx06 should be fixed now, was because `toNum()` was called on the state, and since state was of type `any`, it didn't throw type safety errors. 

I have fixed the state, by changing it to back to string (HassEntity type).

File: src/index.ts [hunk 7]
```diff
@@ -635,21 +574,21 @@ export class SunsynkPowerFlowCard extends LitElement {
         let formattedResultTime = '';
         let duration = '';
 
-        const battenergy = this.hass.states[config.battery.energy] || {state: config.battery.energy ?? ''};
-        let battery_energy = Utils.toNum(battenergy.state, 0);
-        if (battery_voltage && state_battery_rated_capacity) {
-            battery_energy = Utils.toNum(battery_voltage * state_battery_rated_capacity, 0)
+        const battenergy = this.getEntity('battery.energy', {state: config.battery.energy.toString() ?? ''});
+        let battery_energy = battenergy.toNum(0);
+        if (battery_voltage && state_battery_rated_capacity.notEmpty()) {
```
Comment (slipx06) [inline src/index.ts:579]: check if state is empty

**Comment by slipx06** [comment]
I think this is good to merge now.

**Comment by Pho3niX90** [comment]
That should fix it. The optional chaining operator fixes these issues `?`, it will then only call toString() if the object is not null.

**Pull request merged**